### PR TITLE
Fix paginator for falling back to default sort.

### PR DESCRIPTION
**Pull request by dereuromark**
Resolves https://github.com/cakephp/cakephp/issues/12743

They main issue is around

	/**
	 * @var array
	 */
	public $paginate = ['order' => ['Countries.sort' => 'DESC']];

and

    <?php echo $this->Paginator->sort('sort') ?>

where the aliasing silently makes this bug appear right now.

Check out live bug @ https://sandbox.dereuromark.de/data/countries
Click on the "sort" filter, see the URL generated, but never applied.
This has been in Paginator since 3.5 or 3.6 very early it seems (probably around Paginator extract).

Paginator fixes the actual bug around the query adjustment.
The PaginationHelper fix then re-adjusts the URLs for the primary model and non-aliases default columns used.
So that with default order "sort DESC" the URLs generated and clicked are:
- `/countries`
- `/countries?sort=sort&direction=asc`
- `/countries` again

Any possible side effects here?
How can we test this, as the tests do not include the options, only params?

Additional adjustment:
directionDefault is now also consolidated to lowercase version when passed on.

### CODE REVIEW

File: src/View/Helper/PaginatorHelper.php [hunk 1]
```diff
@@ -524,6 +524,13 @@ public function generateUrlParams(array $options = [], $model = null)
         $paging = $this->params($model);
         $paging += ['page' => null, 'sort' => null, 'direction' => null, 'limit' => null];
 
+        if (!empty($paging['sort']) && !empty($options['sort']) && strpos($options['sort'], '.') === false) {
+            $paging['sort'] = $this->_removeAlias($paging['sort'], $model = null);
```
Comment (markstory) [inline src/View/Helper/PaginatorHelper.php]: You're assigning to `model` here which creates a side-effect inside a method call.

File: src/Datasource/Paginator.php [hunk 2]
```diff
@@ -401,6 +406,35 @@ public function validateSort(RepositoryInterface $object, array $options)
...
+        $result = [];
+        foreach ($fields as $field => $sort) {
+            if (strpos($field, '.') === false) {
+                $result[$field] = $sort;
+                continue;
+            }
+
+            list ($alias, $currentField) = explode('.', $field);
+
+            if ($alias !== $model) {
```
Comment (markstory) [inline src/Datasource/Paginator.php]: Why is this version `!==` but the one in the helper uses `===`?

File: tests/TestCase/Datasource/PaginatorTest.php [hunk 3]
```diff
@@ -321,7 +321,7 @@ public function testDefaultPaginateParamsIntoRequest()
         $this->Paginator->paginate($table, [], $settings);
         $pagingParams = $this->Paginator->getPagingParams();
         $this->assertEquals('PaginatorPosts.id', $pagingParams['PaginatorPosts']['sortDefault']);
-        $this->assertEquals('DESC', $pagingParams['PaginatorPosts']['directionDefault']);
+        $this->assertEquals('desc', $pagingParams['PaginatorPosts']['directionDefault']);
```
Comment (markstory) [inline tests/TestCase/Datasource/PaginatorTest.php]: This could break userland code. It isn't very likely but it could happen.

**Review by markstory — COMMENTED** [review]
Why does your paginate order value and HTML sort input use different key/column names? While we could infer which model name like this, it is simpler for us and less magical for users if all the keys used need to match.

File: src/Datasource/Paginator.php (hunk 2 above)
Comment (dereuromark) [reply src/Datasource/Paginator.php]: This was just a bail early here, I can invert the logic to be in sync.

**Comment by markstory** [comment]
> The main issue is the silent breaking here.

Sort conditions are supposed to silently fail when they don't match up.

This appears to be caused by the `order` key containing `['sort' => 'asc', 'Countries.sort' => 'desc']`  after the default `order` and `sort` data are merged. Then the `_prefix()` method is preferring the prefixed key. Perhaps the `_prefix` method should not overwrite keys if they already exist?

**Comment by markstory** [comment]
Are you certain that the behavior you want has ever worked? I can take a look at 3.5, but I get the feeling that this ever worked based on how extensive the changes required to fix this problem are.

**Comment by dereuromark** [comment]
Not fully sure. But that would mean that many apps have this problem by design for no reason IMO.
This shimming is not expensive given a normal dispatcher cycle for pagination, and since the scope is limited to only this specific use case also does less harm having than not having.

Once we can revisit pagination into a common field for sortfield+direction all these issues will then hopefully resolve themselves.

**Comment by dereuromark** [comment]
They do match, the incoming query strings just are not prefixed as per default bake template.
Thing is, the order in the controllers usually have to be model prefixed as otherwise the SQL error about unclear field is thrown (if the same field exists in the contained belongs to relations, which happens a lot).
So this here is really useful for the primary model to work flawless and without painfully upgrading all templates here.

I understand your caution, but let's maybe just check that the scope of the change is indeed as limited as I suspect. This way it will work better also for the default templates and existing default orders.
Check the condition about removing: Only if also the incoming one is not prefixe (same model), and also only without whitelist afaik. Otherwise all remains the same.

> Do you mean combining sort/direction into a single query parameter like sort=-created?

Yes, please see the discussion https://github.com/cakephp/cakephp/issues/10028#issuecomment-272812357
Having a common string generated for possible orders and used by both helper and component, you can easily then build all use cases, incl complex multi-sort, without all the overhead we currently have due to the n*m matrix like options. But a discussion for a future implementation.

**Comment by dereuromark** [comment]
@markstory Do you rather want to put this into 3.7 (RC) then?

**Pull request merged**